File: Seminar12/src/main/java/org/hse/software/construction/restapi/controller/UserController.java
package org.hse.software.construction.restapi.controller;

import java.util.List;
import java.util.UUID;
import org.hse.software.construction.restapi.model.User;
import org.hse.software.construction.restapi.service.UserService;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
public class UserController {

  private final UserService userService;

  public UserController(@Qualifier("defaultUserService") UserService userService) {
    this.userService = userService;
  }

  @PostMapping("user")
  public User createUser(@RequestBody User user) {
    return userService.createUser(user);
  }

  @GetMapping("/user/{id}")
  public User getUserById(@PathVariable String id) {
    return userService.getUserById(UUID.fromString(id));
  }

  @GetMapping("/user")
  public List<User> getUserByName(@RequestParam String name) {
    return userService.getUserByName(name);
  }
}

File: GraphQl/src/main/kotlin/org/hse/graphQl/controller/UserController.kt
package org.hse.graphQl.controller

import org.hse.graphQl.model.User
import org.springframework.graphql.data.method.annotation.Argument
import org.springframework.graphql.data.method.annotation.MutationMapping
import org.springframework.graphql.data.method.annotation.QueryMapping
import org.springframework.stereotype.Controller
import java.util.*

@Controller
class UserController {

    private val users: MutableMap<String, User> = mutableMapOf()

    @QueryMapping
    fun findUserById(@Argument id: String): User? = users[id]

    @QueryMapping
    fun findUsersByName(@Argument name: String): List<User> = users
            .values
            .filter { it.name == name }

    @QueryMapping
    fun findAllUsers(): List<User> = users.values.toList()

    @MutationMapping
    fun createUser(@Argument name: String, @Argument email: String?): User {
        val userId = UUID.randomUUID().toString()
        return User(
                id = userId,
                name = name,
                email = email
        ).also {
            users[userId] = it
        }
    }
}